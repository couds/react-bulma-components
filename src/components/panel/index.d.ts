import { BulmaComponent } from '..';
import { Color } from '..';

interface PanelProps {
  color?: Color;
}

interface PanelTabProps {
  active?: boolean;
}

interface PanelBlockProps {
  active?: boolean;
}

declare const Panel: BulmaComponent<PanelProps, 'nav'> & {
  Header: BulmaComponent<{}, 'div'>;
  Tabs: BulmaComponent<{}, 'div'> & {
    Tab: BulmaComponent<PanelTabProps, 'a'>;
  };
  Block: BulmaComponent<PanelBlockProps, 'div'>;
  Icon: BulmaComponent<{}, 'span'>;
};

export default Panel;
