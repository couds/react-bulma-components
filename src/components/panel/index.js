import './panel.sass';

export * from './panel';
