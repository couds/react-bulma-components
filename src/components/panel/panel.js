import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Block from './components/block';
import Header from './components/header';
import Icon from './components/icon';
import Tabs from './components/tabs';
import modifiers from '../../modifiers';
import Element from '../element';
import renderAsShape from '../../modifiers/render-as';
import CONSTANTS from '../../constants';

const Panel = ({ color, className, ...props }) => (
  <Element
    {...props}
    className={classnames('panel', className, {
      [`is-${color}`]: color,
    })}
  />
);

Panel.Header = Header;

Panel.Tabs = Tabs;

Panel.Block = Block;

Panel.Icon = Icon;

Panel.propTypes = {
  ...modifiers.propTypes,
  className: PropTypes.string,
  renderAs: renderAsShape,
  color: PropTypes.oneOf([null, '', ...Object.values(CONSTANTS.COLORS)]),
};

Panel.defaultProps = {
  ...modifiers.defaultProps,
  className: undefined,
  renderAs: 'nav',
  color: undefined,
};

export default Panel;
