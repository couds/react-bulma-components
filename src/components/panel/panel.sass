@import '../../utils/_all.sass'
@import '~bulma/sass/components/panel.sass'
