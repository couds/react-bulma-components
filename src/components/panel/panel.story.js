import React from 'react';

import { storiesOf } from '@storybook/react';
import Panel from 'react-bulma-components/lib/components/panel';
import {
  Control,
  Input,
  Checkbox,
} from 'react-bulma-components/lib/components/form';
import Icon from 'react-bulma-components/lib/components/icon';
import Button from 'react-bulma-components/lib/components/button';
import { select } from '@storybook/addon-knobs';

import CONSTANTS from '../../constants';

storiesOf('Panel', module)
  .addDecorator(story => <div style={{ margin: 10 }}>{story()}</div>)
  .add('Default', () => (
    <Panel color={select('color', { ...CONSTANTS.COLORS, default: '' })}>
      <Panel.Header>repositories</Panel.Header>
      <Panel.Block>
        <Control>
          <Input size="small" type="text" placeholder="search" />
        </Control>
      </Panel.Block>
      <Panel.Tabs className="panel-tabs">
        <Panel.Tabs.Tab active>all</Panel.Tabs.Tab>
        <Panel.Tabs.Tab>public</Panel.Tabs.Tab>
        <Panel.Tabs.Tab>private</Panel.Tabs.Tab>
        <Panel.Tabs.Tab>sources</Panel.Tabs.Tab>
        <Panel.Tabs.Tab>forks</Panel.Tabs.Tab>
      </Panel.Tabs>
      <Panel.Block renderAs="a" active>
        <Panel.Icon renderAs={Icon} icon="angle-down" />
        bulma
      </Panel.Block>
      <Panel.Block>
        <Panel.Icon renderAs={Icon} icon="angle-down" />
        react-bulma-components
      </Panel.Block>
      <Panel.Block>
        <Panel.Icon renderAs={Icon} icon="angle-down" />
        minireset.css
      </Panel.Block>
      <Panel.Block>
        <Panel.Icon renderAs={Icon} icon="angle-down" />
        jgthms.github.io
      </Panel.Block>
      <Panel.Block>
        <Panel.Icon renderAs={Icon} icon="angle-down" />
        couds.gidhub.io
      </Panel.Block>
      <Panel.Block>
        <Panel.Icon renderAs={Icon} icon="angle-down" />
        mojs
      </Panel.Block>
      <Panel.Block renderAs="label" className="panel-block">
        <Checkbox />
        remember me
      </Panel.Block>
      <Panel.Block>
        <Button fullwidth color="link" outlined>
          reset all filters
        </Button>
      </Panel.Block>
    </Panel>
  ));
