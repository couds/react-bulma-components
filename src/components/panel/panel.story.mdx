import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs/blocks';
import { Panel } from '../..';
import * as stories from './panel.story';
import CommonProps from '../../../.storybook/common-props';

<Meta title="Components/Panel" />

# Panel

A composable panel, for compact controls

The panel is a container for several types:

- `Panel.Header` as the first child
- `Panel.Tabs` for navigation
- `Panel.Block` which can contain other elements, like:
  - `Form.Control`
  - `Form.Input`
  - `Button`
  - `Panel.Icon`

The `Panel.Block` can be an anchor tag `<a>` or a label `<label>` with a checkbox inside.

## Props

<ArgsTable of={Panel} />

## Components

### `Panel.Header`

<ArgsTable of={Panel.Header} />

### `Panel.Tabs`

<ArgsTable of={Panel.Tabs} />

### `Panel.Tabs.Tab`

<ArgsTable of={Panel.Tabs.Tab} />

### `Panel.Block`

<ArgsTable of={Panel.Block} />

### `Panel.Icon`

<ArgsTable of={Panel.Icon} />

<CommonProps />

## Example

<Canvas>
  <Story story={stories.Default} name="Panel" />
</Canvas>

## Related

- [Official documentation](https://bulma.io/documentation/components/panel/)