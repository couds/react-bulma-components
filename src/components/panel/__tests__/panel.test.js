import React from 'react';
import { render } from 'react-testing-library';
import { Panel } from '../panel';

describe('Panel component', () => {
  it.each([[Panel], [Panel.Block], [Panel.Header], [Panel.Icon], [Panel.Tabs], [Panel.Tabs.Tab]])(
    'should render',
    Component => {
      const { asFragment } = render(<Component />);
      expect(asFragment()).toMatchSnapshot();
    }
  );
  it('should render with all', () => {
    const { asFragment } = render(
      <Panel>
        <Panel.Header>repositories</Panel.Header>
        <Panel.Block>
          <div>Control</div>
        </Panel.Block>
        <Panel.Tabs className="panel-tabs">
          <Panel.Tabs.Tab active>all</Panel.Tabs.Tab>
          <Panel.Tabs.Tab>public</Panel.Tabs.Tab>
          <Panel.Tabs.Tab>private</Panel.Tabs.Tab>
          <Panel.Tabs.Tab>sources</Panel.Tabs.Tab>
          <Panel.Tabs.Tab>forks</Panel.Tabs.Tab>
        </Panel.Tabs>
        <Panel.Block renderAs="a" active>
          <Panel.Icon>
            <i className="fa fa-bars" />
          </Panel.Icon>
          bulma
        </Panel.Block>
        <Panel.Block renderAs="label" className="panel-block">
          <input type="checkbox" />
          remember me
        </Panel.Block>
        <Panel.Block>
          <button type="button">reset all filters</button>
        </Panel.Block>
      </Panel>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
