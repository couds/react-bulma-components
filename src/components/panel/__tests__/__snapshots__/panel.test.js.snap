// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel component should render 1`] = `
<DocumentFragment>
  <nav
    class="panel"
  />
</DocumentFragment>
`;

exports[`Panel component should render 2`] = `
<DocumentFragment>
  <div
    class="panel-block"
  />
</DocumentFragment>
`;

exports[`Panel component should render 3`] = `
<DocumentFragment>
  <div
    class="panel-heading"
  />
</DocumentFragment>
`;

exports[`Panel component should render 4`] = `
<DocumentFragment>
  <span
    class="panel-icon"
  />
</DocumentFragment>
`;

exports[`Panel component should render 5`] = `
<DocumentFragment>
  <div
    class="panel-tabs"
  />
</DocumentFragment>
`;

exports[`Panel component should render 6`] = `
<DocumentFragment>
  <a />
</DocumentFragment>
`;

exports[`Panel component should render with all 1`] = `
<DocumentFragment>
  <nav
    class="panel"
  >
    <div
      class="panel-heading"
    >
      repositories
    </div>
    <div
      class="panel-block"
    >
      <div>
        Control
      </div>
    </div>
    <div
      class="panel-tabs panel-tabs"
    >
      <a
        class="is-active"
      >
        all
      </a>
      <a>
        public
      </a>
      <a>
        private
      </a>
      <a>
        sources
      </a>
      <a>
        forks
      </a>
    </div>
    <a
      class="panel-block is-active"
    >
      <span
        class="panel-icon"
      >
        <i
          class="fa fa-bars"
        />
      </span>
      bulma
    </a>
    <label
      class="panel-block panel-block"
    >
      <input
        type="checkbox"
      />
      remember me
    </label>
    <div
      class="panel-block"
    >
      <button
        type="button"
      >
        reset all filters
      </button>
    </div>
  </nav>
</DocumentFragment>
`;
